nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,256 @@
+                    f"[{name} LOG] {msg}", {"machine": name})
+
+            def _log_machinestate(msg):
+                return self.log.log_machinestate(
+                    f"[{name} MCS] {msg}", {"machine": name})
+
+            args = {
+                "startCommand": command,
+                "keepVmState": keep_vm_state,
+                "name": name,
+                "redirectSerial": os.environ.get("USE_SERIAL", "0") == "1",
+                "log_serial": _log_serial,
+                "log_machinestate": _log_machinestate,
+                "tmp_dir": tmp_dir,
+            }
+            return self.machine_class(machine_config_modifier(args))
+
+        self.machines = [create_machine(s) for s in self.vm_scripts]
+
+        @atexit.register
+        def clean_up() -> None:
+            with self.log.nested("cleaning up"):
+                for machine in machines:
+                    machine.release()
+                for _, _, process, _ in vde_sockets:
+                    process.terminate()
+            self.log.release()
+
+    @contextmanager
+    def subtest(self, name: str) -> Iterator[None]:
+        with self.log.nested(name):
+            try:
+                yield
+                return True
+            except Exception as e:
+                self.log.log(f'Test "{name}" failed with error: "{e}"')
+                raise e
+
+        return False
+
+    def export_symbols(self):
+        global machines
+        machines = self.machines
+        machine_eval = [
+            "global {0}; {0} = machines[{1}]".format(m.name, idx) for idx, m in enumerate(machines)
+        ]
+
+        exec("\n".join(machine_eval))
+
+        global start_all, test_script
+        start_all = self.start_all
+
+        global subtest
+
+        def subtest(name):
+            self.subtest(name)
+
+    def run_tests(self) -> None:
+        tests = os.environ.get("tests", None)
+        if tests is not None:
+            with self.log.nested("running the VM test script"):
+                try:
+                    exec(tests, globals())
+                except Exception:
+                    common.eprint("error: ")
+                    traceback.print_exc()
+                    sys.exit(1)
+        else:
+            ptpython.repl.embed(
+                locals(), globals(),
+                configure=self.configure_python_repl)
+
+        # TODO: Collect coverage data
+
+        for machine in self.machines:
+            if machine.is_up():
+                machine.execute("sync")
+
+    def start_all(self) -> None:
+        with self.log.nested("starting all VMs"):
+            for machine in self.machines:
+                machine.start()
+
+    def join_all(self) -> None:
+        with self.log.nested("waiting for all VMs to finish"):
+            for machine in self.machines:
+                machine.wait_for_shutdown()
+
+    def create_vlan(self, vlan_nr: str) -> Tuple[str, str, "subprocess.Popen[bytes]", Any]:
```

Comment:
it seems to me like this function being a member method of `Driver` only makes logging easier.
As logging is done only at the beginning, i suggest making this a free function and emit the log line at the caller site.

Reply:
There are other such cases, I'll review them, too and decide case by case.

Reply:
`VLan` is now a proper object, just as `StartCommand` and `Machine`. :smile: